pydantic/fields.py

```diff
@@ -1107,7 +1107,10 @@ def _validate_discriminated_union(
         assert self.discriminator_alias is not None
 
         try:
-            discriminator_value = v[self.discriminator_alias]
+            try:
+                discriminator_value = v[self.discriminator_alias]
+            except KeyError:
+                discriminator_value = v[self.discriminator_key]
```

Comment:
Could you check the value of `cls.Config. allow_population_by_field_name` or similar before attempting to use this attribute?

In this case, you may need to rework the way the try-excepts work to make sure it still hits the KeyError even when population by field name is not allowed.

Reply:
Makes sense, will do!

Reply:
@dmontagu I simplified my test models (no alias generator) and made more exhaustive tests.

It looks like there's no need to check `cls.Config.allow_population_by_field`. 

Looks like that's already taken care of by how the values are populated: https://github.com/pydantic/pydantic/blob/1.10.X-fixes/pydantic/main.py#L1057-L1059

Reply:
@benwah this is an exotic example, but it demonstrates why the config check is still necessary:
```python
from pydantic import BaseModel
from typing import Annotated, Literal

from pydantic import Field


def to_upper(field_name: str) -> str:
    return field_name.upper()


class PyBase(BaseModel):
    class Config:
        pass


class Dog(PyBase):
    x: Literal["dog"] = Field("dog", alias='y')
    y: None = Field(None, alias='x')


class Cat(PyBase):
    x: Literal["cat"] = Field("cat", alias='y')
    y: None = Field(None, alias='x')


Pet = Annotated[Cat | Dog, Field(discriminator="x")]


class Model(PyBase):
    pet: Pet

m = Model.parse_obj({'pet': {'x': 'abc'}})  # note: due to aliases, x is NOT the discriminator
# On main:
"""
pydantic.error_wrappers.ValidationError: 1 validation error for Model
pet
  Discriminator 'x' is missing in value
"""

# On your branch:
"""
pydantic.error_wrappers.ValidationError: 1 validation error for Model
pet
  No match for discriminator 'x' and value 'abc' (allowed values: 'cat', 'dog')
"""
```
While the error is admittedly confusing either way (always a possibility when you have aliases that are actually names of other fields..), I would argue the error in this scenario is more correct on main than not.

Considering `model` is present in this function call I think it should be reasonable to check the value of the config setting and eliminate this (or other possible issues I haven't considered) as a potential regression.

Reply:
I've implemented the changes you've suggested, but I'm wondering it now makes sense that this call should fail:

https://github.com/pydantic/pydantic/pull/5736/commits/b17a4c95d49ec2e878f6a298b1e4db448c09ec1a#diff-e2a21a51ea103c68ad6398fa0acb2c702bd756957159ab9498354e8227cdc5bbL308-R319

The discriminator is owned by the sub-model `B`, which should accept populating by field name.

However now we consider that the discriminator is missing.

Also shouldn't the `loc` be `("s", "B", "lit")`?

Reply:
.. also please consider this change in error message:

https://github.com/pydantic/pydantic/commit/b17a4c95d49ec2e878f6a298b1e4db448c09ec1a#diff-e2a21a51ea103c68ad6398fa0acb2c702bd756957159ab9498354e8227cdc5bbL313-R330

Reply:
I see your point; in v2 we go by the property name rather than the alias when specifying the discriminator so, although it is perhaps even a bit _more_ restrictive, this particular inconsistency has been eliminated.

I see arguments in favor of each approach here, but my gut reaction is that I'd be more comfortable being more restrictive rather than less (so would prefer to require the config setting be set). If that is enough to solve your problem, let's go with that. (I.e., with the branch as it currently stands).

Reply:
Sounds good! This is indeed enough to fix my problem.